### PR TITLE
Fixed #31548 -- Fixed URLValidator crash on non-strings.

### PR DESCRIPTION
**Pull request by yash-saini275**

### CODE REVIEW

File: django/core/validators.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def __init__(self, schemes=None, **kwargs):
 
     def __call__(self, value):
         # Check first if the scheme is valid
-        scheme = value.split('://')[0].lower()
+        scheme = str(value).split('://')[0].lower()
```
Comment (felixxm) [inline django/core/validators.py:101]: I think we should be more explicit here, maybe:
```python
if not value or '://' not in value:
    raise ValidationError(self.message, code=self.code)
...
```

File: tests/validators/tests.py [hunk 2]
```diff
@@ -222,6 +222,9 @@
     (URLValidator(EXTENDED_SCHEMES), 'git+ssh://git@github.com/example/hg-git.git', None),
 
     (URLValidator(EXTENDED_SCHEMES), 'git://-invalid.com', ValidationError),
+
+    (URLValidator(), None, ValidationError),
+    (URLValidator(), 56, ValidationError),
```
Comment (felixxm) [inline tests/validators/tests.py:227]: Can we add test for value without `://`, e.g. `'x'`.

File: tests/validators/tests.py [hunk 3]
```diff
@@ -222,6 +222,9 @@
     (URLValidator(EXTENDED_SCHEMES), 'git+ssh://git@github.com/example/hg-git.git', None),
 
     (URLValidator(EXTENDED_SCHEMES), 'git://-invalid.com', ValidationError),
+
```
Comment (felixxm) [inline tests/validators/tests.py:225]: Chop blank line.

Comment (yash-saini275) [reply tests/validators/tests.py]: Applied all the changes as you said in the next commit. Please review and let me know if anything else also needs to be changed.

File: django/core/validators.py [hunk 4]
```diff
@@ -97,7 +97,11 @@ def __init__(self, schemes=None, **kwargs):
             self.schemes = schemes
 
     def __call__(self, value):
-        # Check first if the scheme is valid
+        # Check first if the url is string or contains '://'
+        if not value or '://' not in str(value):
```
Comment (felixxm) [inline django/core/validators.py:101]: I changed to `isinstance(value, str)` because it worked properly for an empty string or strings without a scheme.

**Comment by yash-saini275** [comment]
thanks @felixxm

**Pull request merged**